Deprecate legacy URI handling & return valid ASCII-safe URIs (#27556)
gradle/gradle: platforms/documentation/docs/src/docs/userguide/releases/upgrading/upgrading_version_8.adoc

[[changes_8.7]]
== Upgrade from 8.6 and earlier

=== Potential breaking changes

=== Deprecations

[[deprecated_legacy_url_decoding]]
==== Deprecated legacy URL decoding behavior
Prior to Gradle 8.3, Gradle would decode a `CharSequence` given to `link:{groovyDslPath}/org.gradle.api.Project.html#org.gradle.api.Project:uri(java.lang.Object)[Project.uri(Object)]` using an algorithm that accepted invalid URLs and improperly decoded others. 
Gradle now uses the `URI` class to parse and decode URLs, but with a fallback to the legacy behavior in the event of an error.

Starting in Gradle 9.0, the fallback will be removed, and an error will be thrown instead.

To fix a deprecation warning, invalid URLs that require the legacy behavior should be re-encoded to be valid URLs, such as in the following examples:

.Legacy URL Conversions
|===
| Original Input | New Input | Reasoning

| `file:relative/path` | `relative/path` | The `file` scheme does not support relative paths.
| `file:relative/path%21` | `relative/path!` | Without a scheme, the path is taken as-is, without decoding.
| `https://example.com/my folder/` | `https://example.com/my%20folder/` | Spaces are not valid in URLs.

> | `https://example.com/my%%badly%encoded%path` | `https://example.com/my%25%25badly%25encoded%25path` | `%` must be encoded as `%25` in URLs, and no `%`-escapes should be invalid.

Author: No, I don't think so. This applies to `project.file`, `UrlArtifactRepository#setUrl(Object)`, and more.

. Run `gradle wrapper --gradle-version {gradleVersion}` to update the project to {gradleVersion}.
. Try to run the project and debug any errors using the <<troubleshooting.adoc#troubleshooting, Troubleshooting Guide>>.

[[changes_8.7]]
== Upgrade from 8.6 and earlier

=== Potential breaking changes

=== Deprecations

[[deprecated_legacy_url_decoding]]
==== Deprecated legacy URL decoding behavior

> Prior to Gradle 8.3, Gradle would decode a `CharSequence` given to `link:{groovyDslPath}/org.gradle.api.Project.html#org.gradle.api.Project:uri(java.lang.Object)[Project.uri(Object)]`

Reviewer: ```suggestion
Prior to Gradle 8.3, Gradle would decode a `CharSequence` given to `link:{groovyDslPath}/org.gradle.api.Project.html#org.gradle.api.Project:uri(java.lang.Object)[Project.uri(Object)]` using an algorithm that accepted invalid URLs and improperly decoded others. 
```

> using a flawed algorithm that accepted invalid URLs and improperly decoded others. Gradle now uses the `URI` class to parse and decode URLs,

Reviewer: ```suggestion
Gradle now uses the `URI` class to parse and decode URLs, but with a fallback to the legacy behavior in the event of an error.
```

> but with a fallback to the legacy behavior in the event of an error.

Reviewer: ```suggestion
```

> Starting in Gradle 9.0, the fallback will be removed, and the error will be thrown instead.

Reviewer: ```suggestion
Starting in Gradle 9.0, the fallback will be removed, and an error will be thrown instead.
```
